File: 416_Partition Equal Subset Sum.java
// 时间复杂度：O(ns)，其中 n 是 nums 的长度，s 是 nums 的元素和（的一半）。
// 空间复杂度：O(s)

class Solution {
    public boolean canPartition(int[] nums) {
        int target = 0;
        // 1. count sum and determine edge case for odd sum
        for (int n : nums) target += n;
        if (target %2 != 0) return false;
        target /= 2;
        // 2. fill the bag with size 'target'
        int[] dp = new int[target + 1];

        for (int n : nums) {
            for (int j = target; j >= n; j--) {
                dp[j] = Math.max(dp[j], dp[j-n] + n);
                // cut branch
                if (dp[j] == target) return true;
            }
        }

        return (dp[target] == target);
    }
}

File: 416_Partition Equal Subset Sum.py
class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        # Key Insight: 0-1背包 -> check if we can fill the bag of size n (half)
        total = sum(nums)
        if total % 2 != 0: return False

        total >>= 1
        dp = [True] + [False] * total

        for n in nums:
            # fill backwards to avoid repeat adding
            for i in range(total, n-1, -1):
                if not dp[i]: dp[i] = dp[i-n]
        return dp[-1]